relay-server/src/actors/processor.rs

```diff
@@ -344,6 +344,94 @@ fn outcome_from_profile_error(err: relay_profiling::ProfileError) -> Outcome {
     Outcome::Invalid(discard_reason)
 }
 
+fn track_sampling_metrics(
+    project_state: &ProjectState,
+    context: &DynamicSamplingContext,
+    event: &Event,
+) {
+    // We only collect this metric for the root transaction event, so ignore secondary projects.
+    if !project_state.is_matching_key(context.public_key) {
+        return;
+    }
+
+    let transaction_info = match event.transaction_info.value() {
+        Some(info) => info,
+        None => return,
+    };
+
+    let changes = match transaction_info.changes.value() {
+        Some(value) => value.as_slice(),
+        None => return,
+    };
+
+    let last_change = changes.last().and_then(|a| a.value());
+    let source = event.get_transaction_source().as_str();
+    let platform = event.platform.as_str().unwrap_or("other");
+    let sdk_name = event.sdk_name();
+    let sdk_version = event.sdk_version();
+
+    metric!(
+        histogram(RelayHistograms::DynamicSamplingChanges) = changes.len() as u64,
+        source = source,
+        platform = platform,
+        sdk_name = sdk_name,
+        sdk_version = sdk_version,
+    );
+
+    if let Some(&total) = transaction_info.propagations.value() {
+        // If there was no change, there were no propagations that happened with a wrong name.
+        let change = last_change
+            .and_then(|c| c.propagations.value())
+            .map_or(0, |v| *v);
+
+        metric!(
+            histogram(RelayHistograms::DynamicSamplingPropagationCount) = change,
+            source = source,
+            platform = platform,
+            sdk_name = sdk_name,
+            sdk_version = sdk_version,
+        );
+
+        let percentage = (change as f64) / (total as f64);
+        if percentage > 0.0 {
+            metric!(
+                histogram(RelayHistograms::DynamicSamplingPropagationPercentage) = percentage,
+                source = source,
+                platform = platform,
+                sdk_name = sdk_name,
+                sdk_version = sdk_version,
+            );
+        }
+    }
+
+    if let (Some(&start), Some(&change), Some(&end)) = (
+        event.start_timestamp.value(),
+        last_change.and_then(|c| c.timestamp.value()),
+        event.timestamp.value(),
+    ) {
+        let delay_ms = (change - start).num_milliseconds().unsigned_abs();
```

Comment:
If `change < start`, we are dealing with garbage data, so I would not emit a metric in that case.